### PR TITLE
Feature: `merge_state` method

### LINKED ISSUE #2643: Support equivalent of torcheval's `merge_state`
## 🚀 Feature

Support a method equivalent to `torcheval`'s [merge_state](https://github.com/pytorch/torcheval/blob/main/torcheval/metrics/metric.py#L92) to allow explicitly reducing metrics when not used under DDP. This is an updated request from https://github.com/Lightning-AI/torchmetrics/issues/2063.
 
### Motivation

When used within DDP, torchmetrics objects support automatic syncing and reduction across ranks. However, there doesn't seem to be support for reduction outside DDP. This will be a good feature to have because it allows using torchmetrics for distributed evaluation using frameworks other than DDP.

### Pitch

Enabling manual reduction makes `torchmetrics` more widely applicable because it can be used in distributed frameworks other than DDP, such as `ray`.

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Fixes #2643 
Seems most of the logic is already implemented in private methods so it is really just creating a public method + testing.

<details>
  <summary>Before submitting</summary>

- [ ] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to **update the docs**?
- [ ] Did you write any new **necessary tests**?

</details>

<details>
  <summary>PR review</summary>

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

</details>

## Did you have fun?

Make sure you had fun coding 🙃

----
📚 Documentation preview 📚: https://torchmetrics--2786.org.readthedocs.build/en/2786/

### CODE REVIEW

File: src/torchmetrics/metric.py [hunk 1]
```diff
@@ -398,6 +397,60 @@ def _forward_reduce_state_update(self, *args: Any, **kwargs: Any) -> Any:
...
+
+        Example:
+            >>> from torchmetrics.aggregation import SumMetric
+            >>> metric1 = SumMetric()
+            >>> metric2 = SumMetric()
+            >>> metric1.update(1)
+            >>> metric2.update(2)
+            >>> metric1.merge_state(metric2)
+            >>> metric1.compute()
+            tensor(3)
```
Comment (SkafteNicki) [inline src/torchmetrics/metric.py:421]:
```suggestion
            tensor(3.)
```

File: src/torchmetrics/metric.py [hunk 2]
```diff
@@ -398,6 +397,60 @@ def _forward_reduce_state_update(self, *args: Any, **kwargs: Any) -> Any:
...
+
+        Example:
+            >>> from torchmetrics.aggregation import SumMetric
+            >>> metric = SumMetric()
+            >>> metric.update(1)
+            >>> # SumMetric has one state variable called `sum_value`
+            >>> metric.merge_state({"sum_value": torch.tensor(2)})
+            >>> metric.compute()
+            tensor(3)
```
Comment (SkafteNicki) [inline src/torchmetrics/metric.py:430]:
```suggestion
            tensor(3.)
```

**Review by Borda — APPROVED** [review]

**Comment by Borda** [comment]
still:
```
/home/jirka/Workingspace/pt-metrics/src/torchmetrics/metric.py:docstring of torchmetrics.metric.Metric:6: WARNING: Explicit markup ends without a blank line; unexpected unindent.
/home/jirka/Workingspace/pt-metrics/src/torchmetrics/metric.py:docstring of torchmetrics.metric.Metric:7: WARNING: Explicit markup ends without a blank line; unexpected unindent.

**Review by lantiga — APPROVED** [review]
Awesome! Might be worth creating an example on how to use TorchMetrics within a 2D or 3D parallel DTensor-based codebase in the future.
